os/drivers/lcd/st7789.c

```diff
@@ -0,0 +1,920 @@
+
+static int st7789_setorientation(FAR struct lcd_dev_s *dev, unsigned int orientation)
+{
+	FAR struct st7789_dev_s *priv = (FAR struct st7789_dev_s *)dev;
+	lcd_lock(priv->spi);
+	st7789_sendcmd(priv, ST7789_MADCTL);
+	SPI_SELECT(priv->spi, SPIDEV_DISPLAY, true);
+	priv->config->cmddata(1);
+	SPI_SEND(priv->spi, LCD_ST7789_DATA_PREFIX | orientation);
+	SPI_SELECT(priv->spi, SPIDEV_DISPLAY, false);
+	lcd_unlock(priv->spi);
+	return OK;
+}
+
+/****************************************************************************
+ * Name:  st7789_getcontrast
+ *
+ * Description:
+ *   Get the current contrast setting (0-CONFIG_LCD_MAXCONTRAST).
+ *
+ ****************************************************************************/
+
+static int st7789_getcontrast(FAR struct lcd_dev_s *dev)
+{
+	FAR struct st7789_dev_s *priv = (FAR struct st7789_dev_s *)dev;
+	uint8_t buf[2];
+	lcd_lock(priv->spi);
+	st7789_sendcmd(priv, ST7789_RDDISBV);
+	st7789_getdata(priv, buf, 2);
+	lcd_unlock(priv->spi);
+	return buf[0] << 2 || buf[1];
+}
+
+/****************************************************************************
+ * Name:  st7789_setcontrast
+ *
+ * Description:
+ *   Set LCD panel contrast (0-CONFIG_LCD_MAXCONTRAST).
+ *
+ ****************************************************************************/
+
+static int st7789_setcontrast(FAR struct lcd_dev_s *dev, unsigned int contrast)
+{
+	FAR struct st7789_dev_s *priv = (FAR struct st7789_dev_s *)dev;
+	lcd_lock(priv->spi);
+	st7789_sendcmd(priv, ST7789_WRDISBV);
+	st7789_senddata(priv, &contrast, 1);
+	lcd_unlock(priv->spi);
+	return OK;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name:  st7789_initialize
+ *
+ * Description:
+ *   Initialize the ST7789 video hardware.  The initial state of the
+ *   LCD is fully initialized, display memory cleared, and the LCD ready
+ *   to use, but with the power setting at 0 (full off == sleep mode).
+ *
+ * Returned Value:
+ *
+ *   On success, this function returns a reference to the LCD object for
+ *   the specified LCD.  NULL is returned on any failure.
+ *
+ ****************************************************************************/
+
+FAR struct lcd_dev_s *st7789_lcdinitialize(FAR struct spi_dev_s *spi, struct st7789_config_s *config)
+{
+	FAR struct st7789_dev_s *priv = &g_lcddev;
+	/* Initialize the driver data structure */
+	priv->dev.getvideoinfo = st7789_getvideoinfo;
+	priv->dev.getplaneinfo = st7789_getplaneinfo;
+	priv->dev.getpower = st7789_getpower;
+	priv->dev.setpower = st7789_setpower;
+	priv->dev.getcontrast = st7789_getcontrast;
+	priv->dev.setcontrast = st7789_setcontrast;
+	priv->dev.setorientation = st7789_setorientation;
+	priv->dev.init = st7789_init;
```

Comment:
it seems that init called only when LCD initialize. Then we ca remove this from lcd ops, and just keep this private function

Reply:
st7789_init - This can be called as ioctl(INIT) by user if he wants to initialize lcd in between. Should we remove this ioctl as well?

Reply:
I think LCD will be initialized for every boot, and it will be kept as initialized states.
only backlight can be toggled, I mean turn on/off
How do you think?

Reply:
We have not provided any ioctl for backlight on/off. It is just sets on when lcd intialize . Should an ioctl be provided for it?

Reply:
My domain knowledge about lcd & backlight is not good.
But as per my experience, some dimming algorithm need to be handled, so below is necessary.
1. On/Off include time value (ex. turn backlight off after x sec)
2. Dimming level(for Local dimming)

Let's consider more and apply it later, not this PR.

Reply:
ok, so lcd initialization (reset lcd and send initialization commands) is done at boot time. Will there be any need for initialization in between also, means from any app? If not, we will remove init ioctl as well.

Reply:
@edwakuwaku Hello Edward, I have a one question.
When PM change device state to normal mode to power save mode, is each of device driver should initialized again?
You can see st7789_init in this PR, actually it initializes something through the SPI.
And st7789 will keep some state, but when pm change mode to sleep, and device lost all state because of very low voltage level, then PM should enable all device driver again.
How do you think?

Reply:
Hello TaeJun님, as per my understanding, before going power save mode (ie. cpu go to sleep), we will save up the status of the whole system, so when the core wakeup from sleep, we do not need to reinitialize the drivers again. Previously we did some demo on UART, it is working fine (ie. no reinitialization needed, data transfer can still be performed) after wakeup from low power mode~
So, I think it should be the same for SPI, once the development is stable, we will trigger PM to see how to integrate both together
In my opinion, PM is highly dependent to application scenario, and LCDC with SPI interface seems included in the second phase for PM, we can discuss about this further

Reply:
@edwakuwaku Thanks for your answer.
@anjana348 I think we can keep init in this time, and if there's no issue with PM later, then remove it at that time.